src/query_plan/query_planning_traversal.rs

```diff
@@ -86,3 +88,116 @@ struct BestQueryPlanInfo {
     /// The cost of this query plan.
     cost: QueryPlanCost,
 }
+
+impl QueryPlanningTraversal {
+    pub(crate) fn new_dependency_graph(&self) -> FetchDependencyGraph {
+        let root_type = if self.is_top_level && self.has_defers {
+            self.parameters
+                .supergraph_schema
+                .schema()
+                .root_operation(self.root_kind.into())
+                .cloned()
+                // A root operation type has to be an object type
+                .map(|type_name| ObjectTypeDefinitionPosition { type_name }.into())
+        } else {
+            None
+        };
+        FetchDependencyGraph::new(
+            self.parameters.supergraph_schema.clone(),
+            self.parameters.federated_query_graph.clone(),
+            root_type,
+            self.starting_id_generation,
+        )
+    }
+
+    fn updated_dependency_graph(
+        &self,
+        dependency_graph: &mut FetchDependencyGraph,
+        path_tree: &OpPathTree,
+    ) -> Result<(), FederationError> {
+        // TODO: is this right?
+        // In `query-planner-js/src/buildPlan.ts`, `isRootPathTree(tree)` is dependent on `tree`,
+        // but type-Tetris gives us something dependent on `self.root_kind`.
+        let is_root_path_tree = self
+            .parameters
+            .federated_query_graph
+            .root_kinds_to_nodes()?
+            .contains_key(&self.root_kind);
```

Comment:
This isn't quite right, since what we'd want to check is whether the `path_tree.node` is a value of the map given by `root_kinds_to_nodes()` (and not for present/absence of the root kind in the map).

Actually it's probably easier here to just check whether `path_tree.node` is a `QueryGraphNodeType::FederatedRootType`, since that only occurs for nodes in `root_kinds_to_nodes`.